File: src/discan.c
/**********************************************************************
 * 
 * discan.c
 *
 * copyright (c) 2001-6, Karl W Broman
 *
 * last modified Dec, 2006
 * first written Oct, 2001
 *
 *     This program is free software; you can redistribute it and/or
 *     modify it under the terms of the GNU General Public License,
 *     version 3, as published by the Free Software Foundation.
 * 
 *     This program is distributed in the hope that it will be useful,
 *     but without any warranty; without even the implied warranty of
 *     merchantability or fitness for a particular purpose.  See the GNU
 *     General Public License, version 3, for more details.
 * 
 *     A copy of the GNU General Public License, version 3, is available
 *     at http://www.r-project.org/Licenses/GPL-3
 *
 * C functions for the R/qtl package
 *
 * These functions are for performing a genome scan with a 
 * single QTL model
 *
 * Contains: R_discan_mr, discan_mr,
 *           R_discan_im, discan_im
 *  
 **********************************************************************/

#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include <R.h>
#include <Rmath.h>
#include <R_ext/PrtUtil.h>
#include <R_ext/Utils.h>
#include "util.h"
#include "discan.h"

/**********************************************************************
 * 
 * R_discan_mr
 *
 * Wrapper for call from R; reorganizes genotype and result matrix
 * and calls discan_mr.
 * 
 **********************************************************************/

void R_discan_mr(int *n_ind, int *n_pos, int *n_gen,
		    int *geno, int *pheno, double *result)
{
  int **Geno;
  double *means;

  reorg_geno(*n_ind, *n_pos, geno, &Geno);
  allocate_double(*n_gen, &means);

  discan_mr(*n_ind, *n_pos, *n_gen, Geno, pheno, result, means);
}

/**********************************************************************
 * 
 * R_discan_im
 *
 * Wrapper for call from R; reorganizes genotype prob and result matrix
 * and calls discan_im.
 * 
 **********************************************************************/

void R_discan_im(int *n_ind, int *n_pos, int *n_gen, 
		 double *genoprob, int *pheno, double *result, 
		 int *maxit, double *tol)
{
  double ***Genoprob, **work;
  double *means;

  reorg_genoprob(*n_ind, *n_pos, *n_gen, genoprob, &Genoprob);
  allocate_dmatrix(3, *n_gen, &work);
  allocate_double(*n_gen, &means);

  discan_im(*n_ind, *n_pos, *n_gen, Genoprob, pheno, result, 
	    *maxit, *tol, work, means);

}

/**********************************************************************
 * 
 * discan_mr
 *
 * Performs genome scan using marker regression for a dichotomous trait
 * 
 * n_ind        Number of individuals
 *
 * n_pos        Number of marker positions
 *
 * n_gen        Number of different genotypes
 *
 * Geno         Genotype matrix
 *
 * pheno        Phenotype data, as a vector
 *
 * result       Upon return, to contain the log10 likelihoods
 *
 * means        Space for the phenotype means for each genotype
 *
 **********************************************************************/

void discan_mr(int n_ind, int n_pos, int n_gen, int **Geno, 
		  int *pheno, double *result, double *means)
{
  int i, j, k, n, tp, *ng, *np;

  /* number of individuals in each genotype group */
  allocate_int(n_gen, &ng);

  /* number of individuals with phenotype=1 in each genotype group */
  allocate_int(n_gen, &np);

  for(i=0; i<n_pos; i++) {
    R_CheckUserInterrupt(); /* check for ^C */

    result[i] = 0.0; n=tp=0; 
    for(j=0; j< n_gen; j++) {
      ng[j] = np[j] = 0;

      for(k=0; k<n_ind; k++) {
	if(Geno[i][k] == j+1) {
	  if(pheno[k]) {
	    np[j]++;
	    tp++;
	  }
	  n++; ng[j]++;
	}
      }
      
      if(ng[j] > 0) means[j] = (double)np[j]/(double)ng[j];
      else /* no individuals with this genotype */
	means[j] = NA_REAL;
    } /* loop over genotype groups */

    /* calculate LOD score */
    for(j=0; j<n_gen; j++) {
      if(np[j] > 0 && np[j] < ng[j]) 
	result[i] += ((double)np[j]*log10(means[j]) +
			 (double)(ng[j]-np[j])*log10(1.0-means[j]));
    }
    if(tp > 0 && tp < n) 
      result[i] -= ((double)tp*log10((double)tp/(double)n) +
		       (double)(n-tp)*log10((double)(n-tp)/(double)n));

  } /* loop over marker positions */
}

/**********************************************************************
 * 
 * discan_im
 *
 * Performs genome scan using interval mapping.  (The multipoint
 * genotype probabilities have already been calculated in 
 * calc.genoprob)
 * 
 * n_ind        Number of individuals
 *
 * n_pos        Number of marker positions
 *
 * n_gen        Number of different genotypes
 *
 * Genoprob     Array of conditional genotype probabilities
 *
 * pheno        Phenotype data, as a vector
 *
 * result       Upon return, to contain the log10 likelihoods
 *
 * maxit        Maximum number of iterations in the EM algorithm
 *
 * tol          Tolerance for determining convergence in EM
 *
 * work         Workspace of dimension 3 x n_gen
 *
 * means        Space for the phenotype means for each genotype
 *
 **********************************************************************/

void discan_im(int n_ind, int n_pos, int n_gen, double ***Genoprob,
	       int *pheno, double *result, 
	       int maxit, double tol, double **work, double *means)
{
  int i, j, k, s, flag=0;
  double sw;

  for(i=0; i<n_pos; i++) { /* loop over marker positions */

    /* initiate EM */
    for(k=0; k<n_gen; k++) {
      means[k] = sw=0.0;
      for(j=0; j<n_ind; j++) {
	sw += Genoprob[k][i][j]; /* k=gen, i=pos, j=ind */
	if(pheno[j]) means[k] += Genoprob[k][i][j];
      }
      means[k] /= sw;
    }

    for(s=0; s < maxit; s++) { /* EM iterations */
    
      R_CheckUserInterrupt(); /* check for ^C */

      /* copy over current estimates */
      for(k=0; k<n_gen; k++) {
	work[0][k] = means[k]; 
	means[k] = work[1][k] = 0.0;
      }
      
      for(j=0; j<n_ind; j++) { /* loop over individuals */
	/* E-step */
	sw = 0.0;
	for(k=0; k<n_gen; k++) {
	  work[2][k] = Genoprob[k][i][j];
	  if(pheno[j]) work[2][k] *= work[0][k];
	  else work[2][k] *= (1.0-work[0][k]);
	  sw += work[2][k];
	}
	for(k=0; k<n_gen; k++) work[2][k] /= sw;

	/* M-step */
	for(k=0; k<n_gen; k++) {
	  work[1][k] += work[2][k];
	  if(pheno[j]) means[k] += work[2][k];
	}
      }
      
      /* complete M-step */
      for(k=0; k<n_gen; k++) 
	means[k] /= work[1][k];

      /* check for convergence */
      flag = 0;
      for(k=0; k<n_gen; k++) {
	if(fabs(means[k] - work[0][k]) > tol*(fabs(work[0][k])+tol*100.0)) {
	  flag = 1;
	  break;
	}
      }

      if(!flag) break;
    } /* end of EM iterations */

    if(flag) warning("Didn't converge!\n");

    /* calculate log lik */
    result[i] = 0.0;
    for(j=0; j<n_ind; j++) {
      sw = 0.0;
      if(pheno[j]) 
	for(k=0; k<n_gen; k++) sw += Genoprob[k][i][j] * means[k];
      else
	for(k=0; k<n_gen; k++) sw += Genoprob[k][i][j] * (1.0-means[k]);
      result[i] += log10(sw);
    }

  } /* end loop over marker positions */
}



/* end of discan.c */

File: R/discan.R
######################################################################
#
# discan.R
#
# copyright (c) 2001-2015, Karl W Broman
# last modified Oct, 2015
# first written Oct, 2001
#
#     This program is free software; you can redistribute it and/or
#     modify it under the terms of the GNU General Public License,
#     version 3, as published by the Free Software Foundation.
#
#     This program is distributed in the hope that it will be useful,
#     but without any warranty; without even the implied warranty of
#     merchantability or fitness for a particular purpose.  See the GNU
#     General Public License, version 3, for more details.
#
#     A copy of the GNU General Public License, version 3, is available
#     at http://www.r-project.org/Licenses/GPL-3
#
# Part of the R/qtl package
# Contains: discan
#
######################################################################

######################################################################
#
# discan: scan genome, calculating LOD scores with single QTL model
#         for a dichotomous trait
#
######################################################################

discan <-
    function(cross, pheno, method=c("em","hk","mr"),
             addcovar=NULL, intcovar=NULL, maxit=4000, tol=1e-4,
             verbose=FALSE, give.warnings=TRUE, ind.noqtl)
{
    method <- match.arg(method)

    n.ind <- nind(cross)
    n.chr <- nchr(cross)
    type <- class(cross)[1]
    if(is.null(addcovar)) n.addcovar <- 0
    else n.addcovar <- ncol(addcovar)
    if(is.null(intcovar)) n.intcovar <- 0
    else n.intcovar <- ncol(intcovar)

    # individuals with no QTL effect
    if(missing(ind.noqtl)) ind.noqtl <- rep(FALSE, nind(cross))
    else {
        if(!is.logical(ind.noqtl) || length(ind.noqtl) != nind(cross))
            stop("ind.noqtl be a logical vector of length n.ind (", nind(cross), ")")

        if(sum(ind.noqtl) > 1) {
            if(method == "mr") {
                ind.noqtl <- rep(FALSE, nind(cross))
                warning("ind.noqtl ignored for method=", method, ", model=binary")
            }
            else if(is.null(addcovar) && (!is.logical(ind.noqtl) || any(ind.noqtl))) {
                ind.noqtl <- rep(FALSE, nind(cross))
                warning("ind.noqtl ignored when no additive covariates")
            }
        }
    }

    if(method=="mr" && n.addcovar+n.intcovar>0)  {
        if(give.warnings) warning("Covariates ignored with method=\"mr\"; use \"em\" instead")
        n.addcovar <- n.intcovar <- addcovar <- intcovar <- 0
    }

    u <- unique(pheno)
    if(any(u != 0 & u != 1))
        stop("Phenotypes must be either 0 or 1.")

    results <- NULL

    llik0 <- c("A"=NA,"X"=NA)
    nullcoef <- list("A"=NA,"X"=NA)

    for(i in 1:n.chr) {

        chrtype <- class(cross$geno[[i]])
        if(chrtype=="X") {
            sexpgm <- getsex(cross)

            ac <- revisecovar(sexpgm,addcovar)
            if(!is.null(addcovar) && (nd <- attr(ac, "n.dropped")) > 0 && give.warnings)
                warning("Dropped ", nd, " additive covariates on X chromosome.")
            if(length(ac)==0) {
                n.ac <- 0
                ac <- NULL
            }
            else n.ac <- ncol(ac)
            ic <- revisecovar(sexpgm,intcovar)
            if(!is.null(intcovar) && (nd <- attr(ic, "n.dropped")) > 0 && give.warnings)
                warning("Dropped ", nd, " interactive covariates on X chromosome.")
            if(length(ic)==0) {
                n.ic <- 0
                ic <- NULL
            }
            else n.ic <- ncol(ic)
        }
        else {
            sexpgm <- NULL
            ac <- addcovar
            n.ac <- n.addcovar
            ic <- intcovar
            n.ic <- n.intcovar
        }

        # get null log liklihood
        if(is.na(llik0[chrtype])) {
            if(n.ac > 0)
                nullfit <- glm(pheno ~ ac, family=binomial(link="logit"))
            else if(i==1)
                nullfit <- glm(pheno ~ 1, family=binomial(link="logit"))
            fitted <- nullfit$fitted
            nullcoef[[chrtype]] <- nullfit$coef
            llik0[chrtype] <- sum(pheno*log10(fitted) + (1-pheno)*log10(1-fitted))
        }

        # get genotype names
        gen.names <- getgenonames(type,chrtype,"full",sexpgm,attributes(cross))
        n.gen <- length(gen.names)

        # pull out genotype data (mr)
        # or genotype probabilities (em)
        if(method == "mr") {
            newgeno <- cross$geno[[i]]$data
            newgeno[is.na(newgeno)] <- 0

            # discard partially informative genotypes
            if(type=="f2") newgeno[newgeno>3] <- 0
            if(type=="4way") newgeno[newgeno>4] <- 0

            # revise X chromosome genotypes
            if(chrtype=="X" && (type=="bc" || type=="f2"))
                newgeno <- reviseXdata(type, "full", sexpgm, geno=newgeno,
                                       cross.attr=attributes(cross))

            n.pos <- ncol(newgeno)
            map <- cross$geno[[i]]$map
            if(is.matrix(map)) map <- map[1,]

            z <- .C("R_discan_mr",
                    as.integer(n.ind),         # number of individuals
                    as.integer(n.pos),         # number of markers
                    as.integer(n.gen),         # number of possible genotypes
                    as.integer(newgeno),       # genotype data
                    as.integer(pheno),          # phenotype data
                    result=as.double(rep(0,n.pos*(n.gen+1))),
                    PACKAGE="qtl")

        }
        else {
            if(!("prob" %in% names(cross$geno[[i]]))) { # need to run calc.genoprob
                if(give.warnings) warning("First running calc.genoprob.")
                cross <- calc.genoprob(cross)
            }
            genoprob <- cross$geno[[i]]$prob
            n.pos <- ncol(genoprob)

            # revise X chromosome genotypes
            if(chrtype=="X" && (type=="bc" || type=="f2"))
                genoprob <- reviseXdata(type, "full", sexpgm, prob=genoprob,
                                        cross.attr=attributes(cross))

            if("map" %in% names(attributes(cross$geno[[i]]$prob)))
                map <- attr(cross$geno[[i]]$prob,"map")
            else {
                stp <- attr(cross$geno[[i]]$prob, "step")
                oe <- attr(cross$geno[[i]]$prob, "off.end")

                if("stepwidth" %in% names(attributes(cross$geno[[i]]$prob)))
                    stpw <- attr(cross$geno[[i]]$prob, "stepwidth")
                else stpw <- "fixed"
                map <- create.map(cross$geno[[i]]$map,stp,oe,stpw)
            }

            if(is.matrix(map)) map <- map[1,]

            if(method=="hk") {
                z <- .C("R_scanone_hk_binary",
                        as.integer(n.ind),         # number of individuals
                        as.integer(n.pos),         # number of markers
                        as.integer(n.gen),         # number of possible genotypes
                        as.double(genoprob),       # genotype probabilities
                        as.double(ac),
                        as.integer(n.ac),
                        as.double(ic),
                        as.integer(n.ic),
                        as.double(pheno),          # phenotype data
                        result=as.double(rep(0,n.pos)),
                        as.double(tol),
                        as.integer(maxit),
                        as.integer(verbose),
                        as.integer(ind.noqtl),
                        PACKAGE="qtl")
            }
            else if(n.ac + n.ic > 0) {

                start <- rep(nullcoef[[chrtype]][1],n.gen)
                if(n.ac > 0)
                    start <- c(start, nullcoef[[chrtype]][-1])
                if(n.ic > 0)
                    start <- c(start, rep(0, n.ic*(n.gen-1)))

                z <- .C("R_discan_covar",
                        as.integer(n.ind),         # number of individuals
                        as.integer(n.pos),         # number of markers
                        as.integer(n.gen),         # number of possible genotypes
                        as.double(genoprob),       # genotype probabilities
                        as.double(ac),
                        as.integer(n.ac),
                        as.double(ic),
                        as.integer(n.ic),
                        as.integer(pheno),          # phenotype data
                        as.double(start),
                        result=as.double(rep(0,n.pos)),
                        as.integer(maxit),
                        as.double(tol),
                        as.integer(verbose),
                        as.integer(ind.noqtl),
                        PACKAGE="qtl")
            }
            else {
                z <- .C("R_discan_im",
                        as.integer(n.ind),         # number of individuals
                        as.integer(n.pos),         # number of markers
                        as.integer(n.gen),         # number of possible genotypes
                        as.double(genoprob),       # genotype probabilities
                        as.integer(pheno),          # phenotype data
                        result=as.double(rep(0,n.pos)),
                        as.integer(maxit),
                        as.double(tol),
                        PACKAGE="qtl")
            }

        }
        z <- matrix(z$result,nrow=n.pos)

        if(method != "mr") z[,1] <- z[,1] - llik0[chrtype]
        z[is.na(z[,1]),1] <- 0

        z <- z[,1,drop=FALSE]

        colnames(z)[1] <- "lod"

        # get null log10 likelihood for the X chromosome
        adjustX <- FALSE
        if(chrtype=="X") {
            # determine which covariates belong in null hypothesis
            temp <- scanoneXnull(type, sexpgm, cross.attr=attributes(cross))
            adjustX <- temp$adjustX
            parX0 <- temp$parX0+n.ac
            sexpgmcovar <- temp$sexpgmcovar
            sexpgmcovar.alt <- temp$sexpgmcovar.alt

            if(adjustX) { # get LOD-score adjustment
                if(n.ac > 0) {
                    nullfitX <- glm(pheno ~ ac+sexpgmcovar,
                                    family=binomial(link="logit"))
                    parX0 <- lm(pheno~ac+sexpgmcovar)$rank
                }
                else
                    nullfitX <- glm(pheno ~ sexpgmcovar,
                                    family=binomial(link="logit"))
                fittedX <- nullfitX$fitted
                llik0X <- sum(pheno*log10(fittedX) + (1-pheno)*log10(1-fittedX))

                # adjust LOD curve
                z <- z - (llik0X - llik0["X"])
            }

        }

        w <- names(map)
        o <- grep("^loc-*[0-9]+",w)
        if(length(o) > 0) # inter-marker locations cited as "c*.loc*"
            w[o] <- paste("c",names(cross$geno)[i],".",w[o],sep="")

        z <- data.frame(chr=rep(names(cross$geno)[i],length(map)),
                        pos=as.numeric(map), z, stringsAsFactors=TRUE)
        rownames(z) <- w

        results <- rbind(results, z)
    } # loop over chromosomes

    class(results) <- c("scanone","data.frame")
    attr(results,"method") <- method
    attr(results,"type") <- type
    attr(results,"model") <- "binary"
    attr(results,"null.log10.lik") <- llik0["A"]
    if(adjustX)
        attr(results,"null.log10.lik.X") <- llik0X

    results
}

# end of discan.R
